compiler/rustc_parse/src/parser/stmt.rs

```diff
@@ -606,21 +641,72 @@ impl<'a> Parser<'a> {
                                     format!("{} {}", comment_marker, doc_comment_marker),
                                     Applicability::MaybeIncorrect,
                                 );
-                        }
+                            }
+
+                            if self.recover_colon_as_semi() {
+                                // recover_colon_as_semi has already emitted a nicer error.
+                                e.cancel();
```

Comment:
```suggestion
                                e.delay_as_bug();
```

to avoid miscompilations if we somehow break `recover_colon_as_semi`.

Reply:
https://github.com/chenyukang/rust/blob/dbcdce30e88a321a055f4bcf362a56f22b6b3dce/compiler/rustc_parse/src/parser/diagnostics.rs#L1532

`recover_colon_as_semi` will also used here, seems to need more extra changes if we change in that way.